pkg/controller/clusterautoscaler/clusterautoscaler_controller.go

```diff
@@ -0,0 +1,243 @@
+package clusterautoscaler
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	autoscalingv1alpha1 "github.com/openshift/cluster-autoscaler-operator/pkg/apis/autoscaling/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	caImage          = "quay.io/bison/cluster-autoscaler:a554b4f5"
+	criticalPod      = "scheduler.alpha.kubernetes.io/critical-pod"
+	caServiceAccount = "cluster-autoscaler"
+)
+
+// Add creates a new ClusterAutoscaler Controller and adds it to the
+// Manager. The Manager will set fields on the Controller and Start it
+// when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &Reconciler{
+		client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("clusterautoscaler-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource ClusterAutoscaler
+	err = c.Watch(&source.Kind{Type: &autoscalingv1alpha1.ClusterAutoscaler{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to secondary resources owned by a ClusterAutoscaler
+	err = c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &autoscalingv1alpha1.ClusterAutoscaler{},
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &Reconciler{}
+
+// Reconciler reconciles a ClusterAutoscaler object
+type Reconciler struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a ClusterAutoscaler
+// object and makes changes based on the state read and what is in the
+// ClusterAutoscaler.Spec
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Printf("Reconciling ClusterAutoscaler %s/%s\n", request.Namespace, request.Name)
```

Comment:
if I'm reading correctly we are not currently ensuring singleton autoscaler here but scoping the crd instance by namespace/name, so this would allow multiple autoscalers atm, correct?

Reply:
Exactly. Wanted to get this in and do that bit next.